### ISSUE TITLE
Two identical references should always be equal

### ISSUE THREAD

**Issue by weifenluo** [xunit/xunit#2271]
When class implement `IReadOnlyList<T>` to return single item of itself, `Assert.Equal` will crash with stack overflow when comparing of instance of this class. MSTest does not have this problem.

Here is the code:
```
using System;
using System.Collections;
using System.Collections.Generic;
using Xunit;

namespace TestProject1
{
    public class UnitTest1
    {
        private sealed class Field : IReadOnlyList<Field>
        {
            Field IReadOnlyList<Field>.this[int index]
            {
                get
                {
                    if (index != 0)
                        throw new ArgumentOutOfRangeException(nameof(index));

                    return this;
                }
            }

            int IReadOnlyCollection<Field>.Count => 1;

            IEnumerator<Field> IEnumerable<Field>.GetEnumerator()
            {
                yield return this;
            }

            IEnumerator IEnumerable.GetEnumerator()
            {
                yield return this;
            }
        }

        [Fact]
        public void Test1()
        {
            Field x = new Field();

            // This works as expected:
            Assert.Same(x, x);

            // This does not work: The active test run was aborted. Reason: Test host process crashed : Stack overflow.
            Assert.Equal(x, x);
        }
    }
}
```

Full project repo here: https://github.com/weifenluo/XunitAssertEqualBug

**Comment by bradwilson** [comment]
I'm willing to mark this as a bug, but how can you not consider a list when returns itself as an item in the list a classic case of an infinite loop? This feels like a design issue. My temptation "fixing" this bug is to actually call this an error condition; a stack overflow is bad (and not illustrative), but I think any list which contains itself when being checked for deep equality should throw a "you shouldn't do this" exception.

**Comment by weifenluo** [comment]
The code feels like a design issue because it is simplified to demonstrate the bug. The actual design is using `IFields` interface instead of `IReadOnlyList<Field>`:

```
/// <summary>Represents set of Field(s), which is specially optimized for single item set and immutability.</summary>
/// <remarks>
/// <para><see cref="Field"/> class implements <see cref="IFields"/>, so a <see cref="Field"/> instance can represent both
/// the field itself and a single item set of field. This can improve performance by avoiding object creation.</para>
/// <para><see cref="IFields"/> can be sealed as immutable, any change to <see cref="IFields"/> may or may not
/// create a new <see cref="IFields"/> instance. Consumer of <see cref="IFields"/> should always assume it's immutable.</para>
/// <para>Static class <see cref="Fields"/> is provided to manipulate <see cref="IFields"/>.</para>
/// </remarks>
public interface IFields : IReadOnlyList<Field>
{
    /// <summary>Gets a value indicates whether this set is sealed.</summary>
    bool IsSealed { get; }

    /// <summary>Seals this set as immutable.</summary>
    /// <returns>This set.</returns>
    /// <remarks>After calling Seal(), subsequent change to this set will create a new object.</remarks>
    IFields Seal();

    /// <summary>Adds the specified field into this set.</summary>
    /// <param name="value">The field to add.</param>
    /// <returns>A new set if there is any modification to current sealed set; otherwise, the current set.</returns>
    IFields Add(Field value);

    /// <summary>Determines whether this set contains a specified field.</summary>
    /// <param name="value">The field to locate in the set.</param>
    /// <returns><see langword="true"/> if the set contains the specified value; otherwise, <see langword="false"/>.</returns>
    bool Contains(Field value);
}
```

Extension methods are provided to use `IField`:
```
public static class Fields
{
    private class EmptyFields : IFields
    {
        public static EmptyFields Singleton = new EmptyFields();
        private EmptyFields()
        {
        }

        public int Count
        {
            get { return 0; }
        }

        public Field this[int index] => throw new ArgumentOutOfRangeException(nameof(index));

        public bool IsSealed
        {
            get { return true; }
        }

        public bool Contains(Field value)
        {
            value.VerifyNotNull(nameof(value));
            return false;
        }

        public IFields Add(Field value)
        {
            value.VerifyNotNull(nameof(value));
            return value;
        }

        public IFields Seal()
        {
            return this;
        }

        public IEnumerator<Field> GetEnumerator()
        {
            return EmptyEnumerator<Field>.Singleton;
        }

        IEnumerator IEnumerable.GetEnumerator()
        {
            return GetEnumerator();
        }

        public override string ToString()
        {
            return "[]";
        }
    }

    private class ListFields : List<Field>, IFields
    {
        private const int HASHSET_THRESHOLD = 5;
        private const int MIN_LIST_CAPACITY = 4;

        private bool _isSealed;
        private HashSet<Field>? _hashSet;

        public ListFields(Field value1, Field value2)
            : this()
        {
            Debug.Assert(value1 != null && value2 != null && value1 != value2);
            AddField(value1);
            AddField(value2);
        }

        private ListFields()
            : base(Math.Max(MIN_LIST_CAPACITY, HASHSET_THRESHOLD - 1))
        {
        }

        public bool IsSealed
        {
            get { return _isSealed; }
        }

        public IFields Seal()
        {
            _isSealed = true;
            return this;
        }

        bool IFields.Contains(Field value) => ContainsField(value);

        private bool ContainsField(Field value)
        {
            value.VerifyNotNull(nameof(value));
            return _hashSet != null ? _hashSet.Contains(value) : this.Contains(value);
        }

        IFields IFields.Add(Field value)
        {
            return AddField(value);
        }

        private IFields AddField(Field value)
        {
            value.VerifyNotNull(nameof(value));

            if (ContainsField(value))
                return this;

            if (!IsSealed)
            {
                base.Add(value);
                if (_hashSet != null)
                    _hashSet.Add(value);
                else if (Count == HASHSET_THRESHOLD)
                    _hashSet = new HashSet<Field>(this);
                return this;
            }

            if (Count == 0)
                return value;
            else
            {
                var result = new ListFields();
                for (int i = 0; i < Count; i++)
                    result.AddField(this[i]);
                result.AddField(value);
                return result;
            }
        }

        public override string ToString()
        {
            return $"[{CommaDelimitedFields}]";
        }

        private string CommaDelimitedFields => string.Join(", ", this.Select(x => $"'{x}'"));
    }

    /// <summary>
    /// Gets an empty fields set.
    /// </summary>
    public static IFields Empty
    {
        get { return EmptyFields.Singleton; }
    }

    internal static IFields New(Field value1, Field value2)
    {
        Debug.Assert(value1 != null && value2 != null && value1 != value2);
        return new ListFields(value1, value2);
    }

    /// <summary>
    /// Creates a new set of fields.
    /// </summary>
    /// <param name="values">The value of fields.</param>
    /// <returns>The field set.</returns>
    public static IFields New(params Field[] values)
    {
        values.VerifyNotNull(nameof(values));

        if (values.Length == 0)
            return Empty;

        IFields result = values.VerifyNotNull(0, nameof(values));
        for (int i = 1; i < values.Length; i++)
            result = result.Add(values.VerifyNotNull(i, nameof(values)));
        return result;
    }

    private static bool ContainsAll(this IFields source, IFields other)
    {
        for (int i = 0; i < other.Count; i++)
        {
            if (!source.Contains(other[i]))
                return false;
        }
        return true;
    }

    /// <summary>Determines whether the current fields is a proper (strict) subset of the specified fields.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The fields to compare to the current fields.</param>
    /// <returns><see langword="true"/> if the current set is a proper subset of the specified collection; otherwise, <see langword="false" />.</returns>
    public static bool IsProperSubsetOf(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        return source.Count < other.Count && other.ContainsAll(source);
    }

    /// <summary>Determines whether the current fields is a proper (strict) superset of the specified fields.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The fields to compare to the current fields.</param>
    /// <returns><see langword="true"/> if the current set is a proper superset of the specified collection; otherwise, <see langword="false" />.</returns>
    public static bool IsProperSupersetOf(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        return source.Count > other.Count && source.ContainsAll(other);
    }

    /// <summary>Determines whether the current fields is a subset of a specified fields.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The fields to compare to the current fields.</param>
    /// <returns><see langword="true"/> if the current set is a subset of the specified collection; otherwise, <see langword="false" />.</returns>
    public static bool IsSubsetOf(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        return source.Count <= other.Count && other.ContainsAll(source);
    }

    /// <summary>Determines whether the current fields is a superset of a specified fields.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The fields to compare to the current fields.</param>
    /// <returns><see langword="true"/> if the current set is a superset of the specified collection; otherwise, <see langword="false" />.</returns>
    public static bool IsSupersetOf(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        return source.Count >= other.Count && source.ContainsAll(other);
    }

    /// <summary>Determines whether the current fields overlaps with the specified fields.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The fields to compare to the current fields.</param>
    /// <returns><see langword="true"/> if the current set overlaps with the specified fields; otherwise, <see langword="false" />.</returns>
    public static bool Overlaps(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        for (int i = 0; i < source.Count; i++)
        {
            if (other.Contains(source[i]))
                return true;
        }
        return false;
    }

    /// <summary>Determines whether the current fields and the specified fields contain the same elements.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The fields to compare to the current fields.</param>
    /// <returns><see langword="true"/> if the current set and the specified fields contain the same elements; otherwise, <see langword="false" />.</returns>
    public static bool SetEquals(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        return source.Count == other.Count && source.ContainsAll(other);
    }

    /// <summary>Ensures set contain all elements that are present in either the current fields or in the specified fields.</summary>
    /// <param name="source">The current fields.</param>
    /// <param name="other">The collection to add elements from.</param>
    /// <returns>A new set if there is any modification to current set and current set sealed; otherwise, the current set.</returns>
    public static IFields Union(this IFields source, IFields other)
    {
        source.VerifyNotNull(nameof(source));
        other.VerifyNotNull(nameof(other));

        for (int i = 0; i < other.Count; i++)
            source = source.Add(other[i]);
        return source;
    }

    internal static IFields VerifyNotEmpty(this IFields fields, string paramName)
    {
        fields.VerifyNotNull(paramName);
        if (fields.Count == 0)
            throw new ArgumentException("The fields cannot be empty.", paramName);
        return fields;
    }
}
```
So `IFields` can be either a singleton `EmptyFields` object (sealed always), single `Field` object (sealed always), or a `ListFields` object:
```
IFields fields = Fields.Empty; // singleton EmptyFields, sealed
fields = fields.Add(field1) // single field1, sealed
fields = fields.Add(field2).Seal()  // ListFields contains field1 and field2, sealed
```

The fix of this bug is actually very simple: always check `object.ReferenceEquals` before checking for deep equality. Two objects should always consider equal when they're same object. You don't need to modify the logic of deep equality checking.

**Comment by bradwilson** [comment]
Available in v2 `2.8.2-pre.13`
Available in v3 `0.1.1-pre.484`

https://xunit.net/docs/using-ci-builds
